dashboard/app/models/concerns/curriculum/course_audiences.rb

```diff
@@ -0,0 +1,38 @@
+# Model concern for common curriculum audience methods
+# To use, include in a model and call the desired method.
+module Curriculum::CourseAudiences
+  extend ActiveSupport::Concern
+
+  def can_be_instructor?(user)
+    # If unit is in a unit group then decide based on unit group audience
+    return unit_group.can_be_instructor?(user) if is_a?(Script) && unit_group
```

Comment:
Interesting.  Ideally, concerns should not need to call `is_a?` and have logic that depends on the context where the concern is used.  One way to fix that here would be to introduce a method called `has_unit_groups?` that is required for all models that want to use this concern.  Another way would be to factor the code a bit differently so this method takes a user and audience and then have regular instance methods in `unit` and `unit_group` call this method.

Side question -- it seems like we often have logic that takes either a unit or a unit_group.  Do we have some construct that represents this unit-or-unit_group concept?

All that said, this code looks like a lot of other code in our system so I'm fine with this as-is.  Just some food for thought.

Reply:
> Side question -- it seems like we often have logic that takes either a unit or a unit_group. Do we have some construct that represents this unit-or-unit_group concept?

We started trying to use CourseVersion to be either the unit_group or script but have recently decided that it was causing more complicated code and confusion so the longer term plan is to put every unit in a unit_group. @bethanyaconnor is working on the plan for that. 

> Interesting. Ideally, concerns should not need to call `is_a?` and have logic that depends on the context where the concern is used. One way to fix that here would be to introduce a method called `has_unit_groups?` that is required for all models that want to use this concern. Another way would be to factor the code a bit differently so this method takes a user and audience and then have regular instance methods in `unit` and `unit_group` call this method.

Thanks so much for sharing these ideas. It seems like `has_unit_groups?` might be the easiest way forward here if we want to get rid of `is_a?`. I would be interested to hear @davidsbailey thoughts as well.

Reply:
Thinking about it a little more I don't feel like has_unit_groups? is that different from is_a? in this case. I'm leaning toward leaving it with that current check.

Reply:
I'm with James on wanting to get rid of the `is_a?(Script)` checks, but I don't see a better way! I as long as these are limited to an initial check at the beginning of each method directing you to the unit group if there is one, then this is not too bad. again, we could do better if we wanted to invest more heavily here, but we don't necessarily want to invest here because we're trying to go back to this all living only on UnitGroup and not on Script.